### DISCUSSION TITLE
Deploying Storybook in a subdirectory

### DISCUSSION THREAD

**Discussion by imballinst** [storybookjs/storybook#17433] [Help] [answered]
Hello, everyone! I have a question, is there an option to make the `iframe.html` load assets from a particular base URL? For example, I have this repository containing Storybook and Docusaurus: https://github.com/imballinst/test-multisite-gh-pages/tree/main/docs

The Storybook is deployed here: https://imballinst.github.io/test-multisite-gh-pages/storybook/. However, it has an error in the console:

```
Loading module from “https://imballinst.github.io/assets/iframe.a90559e3.js” was blocked because of a disallowed MIME type (“text/html”).
```

I believe this is because in the `iframe.html` it has these lines:

```
<script type="module" crossorigin src="/assets/iframe.a90559e3.js"></script>
<link rel="modulepreload" href="/assets/vendor.e47ed5dd.js">
<link rel="stylesheet" href="/assets/iframe.4b187599.css">
```

I'm using Storybook 6.3 because somehow using the Vite builder breaks for 6.4+ (or maybe there's something that I missed). Any kind of help is appreciated. Thanks!

**Comment by imballinst** [comment, marked as answer]
...actually, I found out the answer right now :sweat_smile: I needed to configure the `viteFinal` in `main.js`. Here's how I found out the answer:

1. From a similar issue, but for Webpack: https://github.com/storybookjs/storybook/issues/12444#issuecomment-750545827. The `publicPath` is the key here, so I tried to find the similar config in Vite.
2. I found out that the field that I need to update is `base` from here: https://vitejs.dev/config/#base
3. And so, I updated my `main.js` to the following:

```js
module.exports = {
  stories: ["../src/**/*.stories.mdx", "../src/**/*.stories.@(js|jsx|ts|tsx)"],
  addons: ["@storybook/addon-links", "@storybook/addon-essentials"],
  framework: "@storybook/react",
  core: {
    builder: "storybook-builder-vite",
  },
  async viteFinal(config, { configType }) {
    config.base = process.env.BASE_PATH || config.base;

    // return the customized config
    return config;
  },
};
```

Finally, I run `BASE_PATH=/test-multisite-gh-pages/storybook/ build-storybook` to build. This way, the local `start-storybook` won't break and it will work for subdirectory deployment.

**Reply by isaiahdaviscom** [reply]
@imballinst, couldn't you **CREATE** a **variable** (e.g., build_path) **AND** **set** that **to** the **destination path** (string)? That way you can just run the "build-storybook" command and avoid errors?

**Reply by imballinst** [reply]
@isaiahdaviscom Did you mean instead of having to use environment variable, we hardcode the path? Sure! That works too, although that might be a bit challenging when we want to scale the base path. For example, for each version of a library we want a Storybook:

```
/v16.14/storybook
/v16.15/storybook
/v16.16/storybook
```

In that case, hardcoding would be a chore, right? Because we need to change it for every release. Instead, if we pick out the environment variable approach, we can just tell the CI to use the latest Git tag and have it prepended to the `BASE_PATH`.

**Reply by isaiahdaviscom** [reply]
I believe I understand. Thanks for breaking this down and sorry for not getting back to you sooner.

**Comment by ostersoft** [comment]
For Storybook 6.4.19 what I had to do is to get some help from linux sed command as I am not much into React. Also when you try to set it up, remember to disable cache in your browser, I was tricked by it when posting the original post here. If it still doesn't fully work for you, check the debugger in your browser and try to identify any calls to / instead of your sub dir, then sed them similarly:

`sed -i "s/\/icons/\/MYSUBDIR\/icons/g" ./dist/storybook/storybook/*iframe.bundle.js && sed -i "s/\/font-awesome\/css/\/MYSUBDIR\/font-awesome\/css/g" ./dist/storybook/storybook/iframe.html && sed -i "s/\/mockServiceWorker.js/\/MYSUBDIR\/mockServiceWorker.js/g" ./dist/storybook/storybook/*iframe.bundle.js `


I added it to the build script in the package.json, before the external server deployment. 

Maybe there is a more civilized way of doing it using config injection.
